### PR TITLE
TS definitions major update

### PR DESCRIPTION
**Pull request by Hotell**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows [our guidelines](https://github.com/skatej/skatej/blob/master/docs/CONTRIBUTING.md#commit)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] Updated TypeScript definitions, if applicable
- [ ] Update docs/migrating.md for breaking changes

**Link to an open issue here**
https://github.com/skatejs/skatejs/issues/1102

**Does this PR introduce a breaking change?** (check one with "x")
```
[x] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

Part of 5.x so TS matches new API's which are breaking -> semver

- [x] remove tests that doesn't make any sense anymore
- [x] Because whole LS/mixin public API was changed it needs  to be addressed as well

### CODE REVIEW

**Comment by treshugart** [comment]
Hey @Hotell just wondering where you're at with this.

**Comment by treshugart** [comment]
Just realised I hadn't submitted the previous review. Just comments. Merge when ready.

File: test/ts/sample-component.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        <div>
+          <h1>app</h1>
+          {h('x-countup', { count: 100, obj: { trucks: 'Independent', deck: 'ZERO' } })}
+        </div>
+      )
+    }
+  }
+)
+
+export type ElmProps = { str: string; arr: any[] }
```
[the remark this answers is not in the document]
Comment (Hotell) [reply test/ts/sample-component.tsx]: Elm just use ELM, javascript is dead ! :D hey dude, long time no see :)

**Comment by Hotell** [comment]
Defs are mostly ready, first off agree on format which we started to discuss here 
-> https://github.com/skatejs/renderer-preact/pull/3#issuecomment-332899826

consistency FTW

**Comment by Hotell** [comment]
@treshugart  Heads up, all done! also fixed prettier and lint-staged so we don't have to reformat everytime

File: package.json [hunk 2]
```diff
@@ -144,6 +145,9 @@
     "setupTestFrameworkScriptFile": "./test/unit/setup-test",
     "testEnvironment": "@skatejs/ssr/jest"
   },
+  "prettier": {
+    "singleQuote": true
```
Comment (Hotell) [inline package.json]: default prettier uses double quotes, skate codebase uses singleQuotes

File: package.json [hunk 3]
```diff
@@ -165,7 +168,7 @@
     "virtual"
   ],
   "lint-staged": {
-    "*.(js|json)": [
+    "*.{ts,tsx,js,json}": [
```
Comment (Hotell) [inline package.json]: previous glob didn't catch files properly. new one is applied on everything code related , even TS files -> consistency FTW

File: src/ts-types/types.d.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  // called after render
+  // rendered?(): void;
+
+  renderer(
+    shadowRoot: Node | Element | ShadowRoot,
+    render: (props?: Mixed) => O
+  ): void;
+}
+
+export interface WithComponent<P = Mixed, S = Mixed, C = Mixed>
```
Comment (Hotell) [inline src/ts-types/types.d.ts]: Typescript is flexible, we can extend types via this DRY pattern so there is no need to reimplement  type definitions within `declare class`

File: test/lib/after-mutations.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
-export default function afterMutations (...fns) {
-  setTimeout(function () {
+export default function afterMutations(...fns) {
```
Comment (Hotell) [inline test/lib/after-mutations.js]: prettier

File: test/lib/fixture.js [hunk 6]
```diff
@@ -1,4 +1,4 @@
-module.exports = function (html, tagName) {
+module.exports = function(html, tagName) {
```
Comment (Hotell) [inline test/lib/fixture.js]: prettier

File: test/lib/support.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 // Determine if ES2015 class statics are inherited. In IE 9 and 10 this is not
 // supported. This is useful for skipping tests that require ES2015 class
 // semantics for statics.
-export function classStaticsInheritance () {
+export function classStaticsInheritance() {
```
Comment (Hotell) [inline test/lib/support.js]: prettier

File: test/ts/no-shadow-root.tsx [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+  withChildren,
+  withUnique,
+  withUpdate,
+  WithComponent,
+  Constructor,
+  CustomElement
+} from 'skatejs';
+
+// custom class definition needed to get Generics support for custom mixin composition
+interface CustomComponentBase<P = {}, S = {}>
```
Comment (Hotell) [inline test/ts/no-shadow-root.tsx]: same technique is used for defining `WithComponent` class which allows us to use generics to provide props/state/context

This is how user can provide their custom Base class strong type annotations for custom mixins composition

**Comment by treshugart** [comment]
@Hotell when you're happy with this I'm happy for you to just merge it. It looks good to me as is. We can polish up later if need be.

**Comment by Hotell** [comment]
kk just need to update changed `shadow` api

**Pull request merged**